src/users_state.rs

```diff
@@ -6,22 +6,52 @@ use serde::{Deserialize, Serialize};
 use serde_with::serde_as;
 use serde_with::DisplayFromStr;
 
-#[derive(Serialize, Deserialize, PartialEq, Debug)]
+#[derive(Serialize, Deserialize, PartialEq, Debug, Default, Clone)]
 pub struct SavedUserState {
     pub available: i64,
     pub friendships: Vec<(usize, usize)>,
+
+    #[serde(skip)]
+    pub friendships_by_user: HashMap<usize, Vec<usize>>,
+}
+
+impl SavedUserState {
+    pub fn init_friendships(&mut self) {
+        self.friendships_by_user = HashMap::new();
+
+        for &(user1, user2) in &self.friendships {
+            let user1_friends = self.friendships_by_user.entry(user1).or_insert(vec![]);
+            user1_friends.push(user2);
+
+            let user2_friends = self.friendships_by_user.entry(user2).or_insert(vec![]);
+            user2_friends.push(user1);
+        }
+    }
+
+    pub fn add_friendship(&mut self, user1: usize, user2: usize) {
+        if self.friendships.contains(&(user1, user2)) || self.friendships.contains(&(user2, user1))
+        {
+            return;
+        }
+
+        self.friendships.push((user1, user2));
```

Comment:
I think the code is clearer with a simple if than having to know that since the users are pushed sorted it implies that one clause can be dropped.

Even in that case, the logic would be harder to read since there's manipulation of the tuple elements